src/self-update/README.md

```diff
@@ -0,0 +1,66 @@
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Run-Time Behaviour
+* Self-Update is provided based on usual thin-egde's Software Management.
+  * A plugin for module type "tedge-update" specifically handles thin-edge self-update.
+* The plugin uses Debian Package Manager APT to upgrade requested thin-edge packages.
+* When executing APT the plugin will disable auto dependency solving.
+  * This is to avoid installing more than the requested packages to reduce risc for potential failures.
+* The plugin shall take all requested packages at once from SM Agent (i.E. Plugin API command "update-list" shall be used).
+* The plugin follows the contract of the Software Management Plugin API, with the exception of the way to report final result to the SM Agent:
+  * As the new started SM Agent will be disconnected from tedge-update plugin's child-process, the plugin will store potential `stdout` and the final exit status to some persistent file.
+  * The SM Agent will pickup that file and generates according update response for the mapper.
+* Potentially the spawned plugin-process need to detouched it-self from parent process SM Agent to avoid any impact when SM Agent process is stopped by APT upgrade procedure.
+
+### Deployment
+* The "tedge-update" plugin must not be contained in any thin-edge package that is updated by that plugin (it cannot overwrite it-self while running).
+  * The "tedge-update" plugin shall be delivered as individual thin-edge package.
+  * That "tedge-update" plugin package can be updated using Software Management as any other Software Module (e.g. with normal "APT" module type).
+
+### Robustness in Packages' installation logic
+* Use of external tools (e.g. grep, sed, ...) shall be avoided in package's installation logic. Instead Rust code/libraries shall be used.
+  * That is to reduce external dependencies and so reduce risc for failures.
+* At very beginning of the installation/removal logic all required/touched external ressources (e.g. config files as "mosquitto.conf", external tools (if any left), ...) shall be checked for availibility/accessibility.
+* If one external ressource is not available installation/removal procedure shall fail before any action was performed.
+
+### Robustness in Self-Update processing
+* When "tedge-update" plugin was started, no other update request for any other plugin must be in progress or started by the SM Agent.
+  * Reason: APT will restart the SM Agent during upgrade, what would result into half-executed update procedures of other parallel running plugins and could lead to corrupted system behaviour.
+  * It would be valid to execute any update requests for other plugins when tedge-update has completely finished. 
+    It would be also valid to reject any other update request and report an error to Cloud while tedge upgrade is running. 
+* The plugin shall download all requested packages before any upgrade starts.
```

Comment:
From where the packages are downloaded?

Reply:
Thanks @makr11st  for pointing that out. That's a mistake in the spec. The behaviour is the same as for all other plugins:
The SM agent will manage the download if any URL is given. Without any URL the package manager is resonsible.

Reply:
This scheme will work today as `tedge` packages will only be distributed as binaries from cloud software repositories with a download URL that the agent can use. But, if we ever publish the `tedge` debian packages to the public apt repository or some other package manager's public repository, this scheme may not work as the agent wouldn't know how to download the installers from "some" public repository. Only the plugin would know how to interact with the repository of the underlying package manager.

That being said, for the current drop, where the tedge installers are guaranteed to be delivered as binaries from the cloud software repo with downloadable URLs, it should be fine.

Reply:
> But, if we ever publish the tedge debian packages to the public apt repository or some other package manager's public repository, this scheme may not work as the agent wouldn't know how to download the installers from "some" public repository. Only the plugin would know how to interact with the repository of the underlying package manager.

Good point. In case no URL is given and so the package manager (here: APT) is responsbible for the downloading as part of the installation, the plugin must instruct APT to: 1st download first all package; then 2nd install all packages. I updated the spec accoringly:


- For module type „tedge“ the plugin must instruct APT to first download all package that are part of the entire update request. Therefore implementation of plugin command update-list calls APT separately for download and for installation.
   - First all packages are downloaded by calling APT with command line option `--download-only`.
   - When all packages were successfully downloaded packages are installed by calling APT again with option `--no-download`.
   - If one or more downloads fails, the plugin will not install any package, but return with error.